core/example/longrange/spme.cpp

```diff
@@ -0,0 +1,490 @@
+                  Ur_part += qiqj * erfc(alpha * scal)/scal;
+                }
+              }
+            }
+          }
+        Ur_part *= 0.5;
+        p(idx) += Ur_part;
+      },
+    Ur);
+  Kokkos::fence();
+
+  
+  // computation reciprocal-space contribution
+ 
+  //First, spread the charges onto the mesh
+  
+  double spacing = meshr(1,0)-meshr(0,0);//how far apart the mesh points are (assumed uniform cubic)   
+  
+  //Current method: Each mesh point loops over *all* particles, and gathers charge to it 
+  //                 according to spline interpolation. 
+  //Alternatives: Looping over all particles, using atomics to scatter charge to mesh points
+  //Also, would be nice to loop only over neighbors - spline is only 2 mesh points away maximum
+  auto spread_q = KOKKOS_LAMBDA( const int idx )
+  {
+     double xdist, ydist, zdist;
+     for ( size_t pidx = 0; pidx < particles.size(); ++pidx )
+     {
+        //x-distance between mesh point and particle
+        xdist = min(min(abs(meshr(idx,0) - r(pidx,0)),
+                     abs(meshr(idx,0) - (r(pidx,0) +  1.0))),
+                     abs(meshr(idx,0) - (r(pidx,0) -  1.0)) );//account for periodic bndry
+        //y-distance between mesh point and particle
+        ydist = min(min(abs(meshr(idx,1) - r(pidx,1)),
+                     abs(meshr(idx,1) - (r(pidx,1) +  1.0))),
+                     abs(meshr(idx,1) - (r(pidx,1) -  1.0)) );//account for periodic bndry
+        //z-distance between mesh point and particle
+        zdist = min(min(abs(meshr(idx,2) - r(pidx,2)),
+                     abs(meshr(idx,2) - (r(pidx,2) +  1.0))),
+                     abs(meshr(idx,2) - (r(pidx,2) -  1.0)) );//account for periodic bndry
+
+        if ( xdist <= 2.0*spacing and ydist <= 2.0*spacing and zdist <= 2.0*spacing ) //more efficient way to do this? Skip it? May be unnecessary.
+        {
+           //add charge to mesh point according to spline
+           meshq(idx) += q(pidx) * TPME::oneDspline(2.0-(xdist/spacing))
+                                 * TPME::oneDspline(2.0-(ydist/spacing)) 
+                                 * TPME::oneDspline(2.0-(zdist/spacing));
+        }
+     }       
+  };
+  Kokkos::parallel_for( Kokkos::RangePolicy<ExecutionSpace>(0,meshsize), spread_q ); 
+  Kokkos::fence();
+  
+  //starttime2 = std::chrono::steady_clock::now();
+
+  //std::cout << "Creating BC array" << std::endl;
+  //Create "B*C" array (called theta in Eqn 4.7 SPME paper by Essman)
+  //Can be done at start of run and stored
+  //"meshwidth" should be number of mesh points along any axis. 
+  int meshwidth = std::round(std::pow(meshsize, 1.0/3.0));//Assuming cubic mesh
+
+  //Calculating the values of the BC array involves first shifting the fractional coords
+  //then compute the B and C arrays as described in the paper
+  //This can be done once at the start of a run if the mesh stays constant
+  #ifdef Cabana_ENABLE_Cuda
```

Comment:
Seeing this we should develop a portable FFT interface in the future that handles this automatically.